### ISSUE TITLE
Feature Request: Copy Definition to Clipboard

### ISSUE THREAD

**Issue by moduloapparent** [birtles/rikaichamp#50]
Love Rikaichamp. It's been a godsend since Rikaisama went defunct on the new FF.
One feature missing from it is the ability to copy a definition from the clipboard so I can paste it into an SRS or other program.

**Comment by birtles** [comment]
Thank you very much!
Yes, I had in mind to do that but no-one seemed to ask for it so I assumed it was low priority. Let me see if I can work on it soon.

**Comment by moduloapparent** [comment]
Thanks Brian! That would be pretty amazing to have it :)

**Comment by picsi** [comment]
That's awesome! This is something I never knew I wanted to have, but now I see how it would make my life so much easier.

**Comment by picsi** [comment]
That's awesome! This is something I never knew I wanted to have, but now I see how it would make my life so much easier.

**Comment by birtles** [comment]
Yep, it's お盆休み here so while I still have work, I hope to get time in the evenings this week to work on it. I need to make keyboard keys customizable and do the "hold a key to activate" feature first, however.

**Comment by birtles** [comment]
The "hold a key to activate" feature has shipped and this is next on my list. I notice rikaikun lets you say how many entries you want to copy. However, I'm guessing that >95%? of the time you really just want the top entry and copying those other entries would just get in the way.

I try to keep Rikaichamp simple so I'd like to add as few configuration options as possible. I'm wondering if we should just make it copy the first entry? That would also obviate the need for rikaikun's "Line ending" setting.

The other setting in rikaikun is "Field separator". What are folks' specific requirements here.

One option is just to use a format similar to the data file, e.g.
```
上場 [じょうじょう] /(n,vs) (1) listing (on the stock exchange, etc.)/taking a company public/(2) presentation (e.g. of a play, opera)/performance/(P)/
```

Perhaps we could pretty it up a little:
```
上場【じょうじょう】(n,vs) listing (on the stock exchange, etc.), taking a company public; presentation (e.g. of a play, opera), performance
```

I wonder if a multi-key shortcut makes sense?

e.g.

| Key | Result |
| --- | --- |
| `ce` (copy entry) | `上場【じょうじょう】(n,vs) listing (on the stock exchange, etc.), taking a company public; presentation (e.g. of a play, opera), performance` |
| `cw` (copy word) | `上場` |
| `cf` (copy fields) | `上場 → じょうじょう → (n,vs) → listing (on the stock exchange, etc.), taking a company public; presentation (e.g. of a play, opera), performance` ( → here represents a Tab) |

Then maybe we could even have `2ce` to mean "Copy the second entry". You can see I'm a vim user.

Please let me know if you have any suggestions!

**Comment by birtles** [comment]
Another thought: I'm not sure how folks are using this, but if they're collecting a whole bunch of words at once we could make it so that some other shortcut key lets you append to the clipboard. e.g. you hover over one word, hit `cf` then hover over another and hit, e.g. `3Cf` to append the third entry to the clipboard.

That's probably overkill though and it suffers the same problem as vim where if on one occasion you fail to use the append version of the command you lose all your previous work.

**Comment by picsi** [comment]
>I'm wondering if we should just make it copy the first entry?

I don't think this is a good idea while bug #26 is haunting us.

**Comment by birtles** [comment]
> I don't think this is a good idea while bug #26 is haunting us.

Good point. But I still find that the top entry is the one I want > 95% of the time, and if this feature adds the ability to select other entries then I don't think it will be problem.

(I do plan to fix #26 but it will take a while longer.)

**Comment by moduloapparent** [comment]
Hey Brian!
Really good questions (which I hadn't thought of in any depth till you raised them).

**Copy behaviour**
- How would supporting only two behaviours work e.g. 1) copy first definition in list or 2) copy all definitions in list. 
This would keep it simple for the 80%+ of times where the first entry is the one that's desired, while allowing the 20% odd to be supported (with minor edits by the user to cut out bits that are unnecessary).

**Visual presentation**
Personally I'd prefer the version that shows the numbers between different definitions as it makes it a touch easier to scan for where one definition ends and the other begins. Another way of course would be to create a line-break  or a full paragraph break between the definitions. (This would probably be neater, but come at the livable downside of taking up more space in a card). 

**Multiple clipboard entries support**
- I personally wouldn't use this as I'd be nervous around needing to mentally track what I've copied without a visual means of telling what's in the clipboard.

Thank you so much for working on this, Brian!

**Comment by moduloapparent** [comment]
Hey Brian!
What I had in mind with the copy all entries key was to try simplify the options. I imagined that if you got a word where definition 3 was the correct definition then unless it was given its own hotkey there'd be no easy way to get it. These will hopefully always be edge cases, so if the default behaviour catches the most common then the secondary behaviour can afford to require a bit more work (ie the user deleting all the non-required definitions manually)

Yes, with you on all your other comments!

**Comment by birtles** [comment]
Oh, that makes sense. The current approach I have in mind is as follows:

1. Press <kbd>c</kbd> once and the first entry is highlighted.
2. If you press <kbd>c</kbd> again the next entry is highlighted such that you can cycle through them one by one. (Not sure what happens in kanji view though).
3. When an entry is highlighted, press <kbd>e</kbd> to copy it, or <kbd>w</kbd> to copy just the word, or <kbd>f</kbd> (<kbd>r</kbd>?) to copy the entry as a tab-delimited string.
4. After that the highlight is cleared (but the pop-up remains showing so that you can copy another entry or keep viewing it).
5. Pressing <kbd>Esc</kbd> also clears the highlight.

So in the common case where you just want to copy the first entry you just type <kbd>c</kbd> <kbd>e</kbd> in rapid succession.

In the case where the entry you want is the third one you'd type <kbd>c</kbd> <kbd>c</kbd> <kbd>c</kbd> <kbd>e</kbd>.

Ideally the pop-up itself would show a little guide after pressing <kbd>c</kbd> once saying, e.g. "Copy: `e` = entry, `w` = word, `f` = fields, `Esc` = cancel"). Also, those keys should probably be configurable for non-English speakers.

(That said, this feature is getting a little bit more involved than I originally anticipated so I'm not sure yet if I should come back to it after working on fixing the dictionary database. Initially I thought I'd knock over a few "easy" features then fix all the database issues but at this rate I'm not sure which will land first.)

**Comment by ambiguousrocket** [comment]
Would like to show my interest in this feature too. I copy and paste definitions into Anki often--Chrome's rikaikun extension does have copy to clipboard and it's the only reason why I use the browser. Text hovering is choppy however and it feels much, **_much_** nicer in rikaichamp.

For my own selfish reasons: I would be happy with a single C keypress copying all definitions visible in the pop-up. I limit the number of definitions to 5 and typically delete the ones which I don't use manually.

Good luck!

**Comment by birtles** [comment]
Thanks for the feedback. As a result I started working on this yesterday. I'll see if I can get something done this week.

**Comment by moduloapparent** [comment]
Thanks Brian! looking forward to it :)

**Comment by birtles** [comment]
Just an update here. I've been working on this pretty steadily for the last 2 weeks. I've had to rewrite fairly large chunks of the add-on but I think most of the hard work is now done. Hopefully should be done in a few days' time.

**Comment by birtles** [comment]
Sorry for the delay, I've been unwell. This is basically done now. I just need to test that there is no perf regression from this.

**Comment by moduloapparent** [comment]
Legend! Thank you so much, Brian!
I'll download the update once it propagates to Firefox (I'm not familiar with how to use the version you've got here on Github).

**Comment by birtles** [comment]
Thank you. I'll hopefully publish a new release tomorrow. Otherwise it will be next week.

**Comment by moduloapparent** [comment]
Brian, thank you so much! I've tested the new version and it's great. Really like the way you implemented the definition selection interaction. It makes it super clear what's happening. Very well done on this!

**Comment by birtles** [comment]
Thanks so much! I'm really glad you like it. Sorry it took so long!

(I've noticed, however, that if you don't have the window focused <kbd>Shift</kbd> works but <kbd>c</kbd> does not. I'm not sure yet if there's anything I can do about that, however.)

**Comment by ambiguousrocket** [comment]
Just tested this myself, really awesome work birtles! I absolutely love how the definition that's being copied is highlighted in the pop-up window.

Is there any chance you can get the clipboard feature to work with input fields? I primarily use the extension by transcribing Japanese text into a shrib.com page, and then hovering over words/copying over definitions when I need to.

**Comment by birtles** [comment]
Ah, good point. Yeah, Rikaichamp's handling of text fields (particularly multi-line `<textarea>`) is pretty horrible at the moment. I need to fairly thoroughly redo that.

Mind filing an issue to remind me?

**Comment by birtles** [comment]
Actually, never mind, issue #20 covers this.

**Comment by ambiguousrocket** [comment]
Cool! I'll be keeping an eye on #20 then.
